examples/blake2f.rs

```diff
@@ -0,0 +1,1427 @@
+                (m_vec_values[s[14]], m_vec_values[s[15]]),
+                (&mut v_mid_va_bit_vec, &mut v_mid_vb_bit_vec),
+                (&mut v_mid_vc_bit_vec, &mut v_mid_vd_bit_vec),
+                (&mut v_xor_d_bit_vec, &mut v_xor_b_bit_vec),
+                (&mut b_bit_vec, &mut b_3bits_vec),
+            );
+
+            let ginputs = GInput {
+                round: F::from(r as u64),
+                v_vec: v_vec_values.iter().map(|&v| F::from(v)).collect(),
+                h_vec: h_vec_values.iter().map(|&v| F::from(v)).collect(),
+                m_vec: m_vec_values.iter().map(|&v| F::from(v)).collect(),
+                v_mid1_vec: v_mid1_vec_values.iter().map(|&v| F::from(v)).collect(),
+                v_mid2_vec: v_mid2_vec_values.iter().map(|&v| F::from(v)).collect(),
+                v_mid3_vec: v_mid3_vec_values.iter().map(|&v| F::from(v)).collect(),
+                v_mid4_vec: v_mid4_vec_values.iter().map(|&v| F::from(v)).collect(),
+                v_mid_va_bit_vec,
+                v_mid_vb_bit_vec,
+                v_mid_vc_bit_vec,
+                v_mid_vd_bit_vec,
+                v_xor_d_bit_vec,
+                v_xor_b_bit_vec,
+                b_bit_vec,
+                b_3bits_vec,
+            };
+            ctx.add(&blake2f_g_setup_vec[r as usize], ginputs);
+            v_vec_values = v_mid4_vec_values.clone();
+        }
+
+        let output_vec_values: Vec<u64> = h_vec_values
+            .iter()
+            .zip(
+                v_vec_values[0..8]
+                    .iter()
+                    .zip(v_vec_values[V_LEN as usize / 2..V_LEN as usize].iter()),
+            )
+            .map(|(h, (v1, v2))| h ^ v1 ^ v2)
+            .collect();
+
+        let final_inputs = FinalInput {
+            round: F::from(values.round as u64),
+            v_vec: v_vec_values.iter().map(|&v| F::from(v)).collect(),
+            h_vec: h_vec_values.iter().map(|&v| F::from(v)).collect(),
+            output_vec: output_vec_values.iter().map(|&v| F::from(v)).collect(),
+            v_split_bit_vec: v_vec_values
+                .iter()
+                .map(|&v| split_value_4bits(v as u128, BITS_NUMBER))
+                .collect(),
+            h_split_bit_vec: h_vec_values
+                .iter()
+                .map(|&v| split_value_4bits(v as u128, BITS_NUMBER))
+                .collect(),
+            v_xor_split_bit_vec: v_vec_values[0..V_LEN as usize / 2]
+                .iter()
+                .zip(v_vec_values[V_LEN as usize / 2..V_LEN as usize].iter())
+                .map(|(&v1, &v2)| split_xor_value(v1, v2))
+                .collect(),
+            final_split_bit_vec: output_vec_values
+                .iter()
+                .map(|&output| split_value_4bits(output as u128, BITS_NUMBER))
+                .collect(),
+        };
+        ctx.add(&blake2f_final_step, final_inputs);
+        // ba80a53f981c4d0d, 6a2797b69f12f6e9, 4c212f14685ac4b7, 4b12bb6fdbffa2d1
```

Comment:
@10to4 have you checked this result with other implementation, even better with the EVM implementation?

Reply:
Yes, the input example comes from the document [eip-152](https://github.com/ethereum/EIPs/blob/6572e92dccb2a581c0082befb953050f75d0ece5/EIPS/eip-152.md#example-usage-in-solidity).